staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go

```diff
@@ -82,5 +82,9 @@ type KMSConfig struct {
 	// +optional
 	CacheSize int `json:"cachesize,omitempty"`
 	// configfile is the path to the configuration file for the named KMS provider.
-	ConfigFile string `json:"configfile"`
+	// +optional
+	ConfigFile string `json:"configfile,omitempty"`
+	// the gRPC server listening address, for example "unix:///var/run/kms-provider.sock".
+	// +optional
+	Endpoint string `json:"endpoint,omitempty"`
```

Comment:
At this point you need validation on the type.  Not all combinations are valid and it needs to be enforced.

Also, I don't see a version on this config. Without a version, we won't be able to easily have structural change.  For instance, later allowing a remote listening address will require more fields that should probably be logically grouped.

Reply:
this struct is nested inside EncryptionConfig (which has a kind/apiVersion, though not a standard apimachinery-defined one)

the ConfigFile referenced by the KMS can be removed, I think... we have no KMS plugins, and the only method we're planning to support is this external call, right?

Reply:
1. There will be only gRPC provider, no other plugins. So I will remove "ConfigFile" and the file "plugins.go".
2. There has been "apiVersion" field in config file, if we also add version for KMS, the config file will be:
``` yaml
kind: EncryptionConfig
apiVersion: v1
resources:
  - resources:
    - secrets
    providers:
    - kms:
        apiVersion: v1
        name: grpc
        endpoint: unix:///var/run/kms-provider.sock
```
It seems strange. If there is no other KMS plugins, we can think the "kms" is a part of the whole config.